### PR TITLE
[FLINK-6198] [cep] Update CEP documentation.

### PR DESCRIPTION
**Pull request by kl0u**
R @tzulitai @dawidwys

### CODE REVIEW

File: docs/dev/libs/cep.md [hunk 1]
```diff
@@ -874,3 +1141,19 @@ val alerts = patternStream.select(createAlert(_)))
...
+old version of your job, *i.e.* without having to re-process your past data.
+
+The changes required are:
+
+1. Change your conditions (the ones in the `where(...)` clause) to extend the `SimpleCondition` class instead of 
+implementing the `FilterFunction` interface.
+
+2. Change your functions provided as arguments to the `select(...)` and `flatSelect(...)` methods to expect a list of
+events associated with each state (`List` in `Java`, `Iterable` in `Scala`). This is because with the addition of
+the looping states, multiple input events can much a single (looping) state.
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: much -> match

Comment (kl0u) [inline docs/dev/libs/cep.md]: hehe good "cats" (catch) :P

File: docs/dev/libs/cep.md [hunk 2]
```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
+
+#### Conditions
+
+At every state, and in order to go from one state to the next, you can specify additional **conditions**. 
+These conditions can be related to:
+ 
+ 1. a [property of the incoming event](#conditions-on-properties), e.g. its value should be larger than 5, 
+ or larger than the average value of the previously accepted events.
+
+ 2. the [contiguity of the matching events](#conditions-on-contiguity), e.g. detect pattern `a,b,c` without
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: I would move this to the combining patterns section. It is easier to reason about after understanding the contiguity between single patterns/states.

File: docs/dev/libs/cep.md [hunk 3]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
-For example, one can define that a pattern should occur within 10 seconds via the `within` method. 
+
+Bear in mind that relaxed contiguity means that only the first succeeding matching event will be matched, while
+non-deterministic relaxed contiguity, multiple matches will be emitted for the same beginning.
+
+Finally, it is also possible to define a temporal constraint for the pattern to be valid.
+For example, you can define that a pattern should occur within 10 seconds via the `pattern.within()` method. 
 Temporal patterns are supported for both [processing and event time]({{site.baseurl}}/dev/event_time.html).
 
 <div class="codetabs" markdown="1">
```
Comment (tzulitai) [inline docs/dev/libs/cep.md]: I wonder if it makes sense to break up this table into 2 parts, one placed under the "Individual States" section, and "Combining States" section, to make it clearer which API calls are for describing the construction of an individual state and which ones are for combining different states.

File: docs/dev/libs/cep.md [hunk 4]
```diff
@@ -659,37 +913,45 @@ val patternStream: PatternStream[Event] = CEP.pattern(input, pattern)
...
 <div data-lang="java" markdown="1">
 The `select` method requires a `PatternSelectFunction` implementation.
 A `PatternSelectFunction` has a `select` method which is called for each matching event sequence.
-It receives a map of string/event pairs of the matched events.
-The string is defined by the name of the state to which the event has been matched.
-The `select` method can return exactly one result.
+It receives a match in the form of `Map<String, List<IN>>` where the key is the name of each state in your pattern 
+sequence and the value is a list of all accepted events for that state (`IN` is the type of your input elements).
```
Comment (tzulitai) [inline docs/dev/libs/cep.md]: Perhaps add information about the ordering of the list of matched events?

File: docs/dev/libs/cep.md [hunk 5]
```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
+
+ 2. the [contiguity of the matching events](#conditions-on-contiguity), e.g. detect pattern `a,b,c` without 
+ non-matching events between any matching ones.
+ 
+The latter refers to "looping" states, i.e. states that can accept more than one event, e.g. the `b+` in `a b+ c`, 
+which searches for one or more `b`'s.
+
+##### Conditions on Properties
+
+Conditions on the event properties can be specified via the `pattern.where()` method. These can be either
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: Conditions on the event properties can be specified via the `pattern.where()` **and `pattern.or()`** method

File: docs/dev/libs/cep.md [hunk 6]
```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
+The latter refers to "looping" states, i.e. states that can accept more than one event, e.g. the `b+` in `a b+ c`, 
+which searches for one or more `b`'s.
+
+##### Conditions on Properties
+
+Conditions on the event properties can be specified via the `pattern.where()` method. These can be either 
+`IterativeCondition`s or `SimpleCondition`s.
+
+**Iterative Conditions:** This is the most general type of conditions. This allows to specify a condition that accepts 
+any subsequent event based on some statistic over a subset of the previously accepted events.
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: based on properties of the previously accepted events or even some statistic over a subset of them.

This way we emphasize that we can take into account for example just the previous event.

**Comment by kl0u** [comment]
Thanks a lot for the thorough review @tzulitai ! I will address your comments and then merge.

File: docs/dev/libs/cep.md [hunk 7]
```diff
@@ -215,12 +303,11 @@ start.subtype(classOf[SubEvent]).where(subEvent => ... /* some condition */)
...
-As it can be seen here, the subtype condition can also be combined with an additional filter condition on the subtype.
-In fact, you can always provide multiple conditions by calling `where` and `subtype` multiple times.
-These conditions will then be combined using the logical AND operator.
+**Combining Conditions:** As shown, the `subtype` condition can be combined with additional conditions. 
+In fact, this holds for every condition. You can arbitrarily combine multiple conditions by sequentially calling 
+`where()`. The final result will be the logical **AND** of the results of the individual conditions.
 
-In order to construct or conditions, one has to call the `or` method with a respective filter function.
-Any existing filter function is then ORed with the given one.
+In order to combine conditions using *OR*, you can call the `or` method, as shown below.
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: bold the OR (*OR* -> **OR**)
`or` -> `or()`

File: docs/dev/libs/cep.md [hunk 8]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
 </div>
 </div>
 
-Next, we can append further states to detect complex patterns.
-We can control the contiguity of two succeeding events to be accepted by the pattern.
+##### Conditions on Contiguity
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: I would make it a higher level section and changed Conditions on Contiguity -> Contiguity.

Also I think this is the place where inner contiguity for looping states should be described.

File: docs/dev/libs/cep.md [hunk 9]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+ 
+ 3. Non-Deterministic Relaxed Contiguity: which further relaxes contiguity by also creating alternative
+ matches which ignore also matching events.
 
-Non-strict contiguity means that other events are allowed to occur in-between two matching events.
-A non-strict contiguity pattern state can be created via the `followedBy` or `followedByAny` method.
+To illustrate the above with an example, a pattern sequence `a+ b` (one or more `a`s followed by a `b`) with 
+input `a1, c, a2, b` will have the following results:
+
+ 1. Strict Contiguity: `a2 b` because there is `c` `a1` and `a2` so `a1` is discarded.
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: there is `c` between `a1` and `a2`

File: docs/dev/libs/cep.md [hunk 10]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+
+ 2. Relaxed Contiguity: `a1 b` and `a1 a2 b`, as `c` will get simply ignored.
+ 
+ 3. Non-Deterministic Relaxed Contiguity: `a1 b`, `a2 b` and `a1 a2 b`.
+
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: For the examples I would use some additional signs, cause just with the formatting it is hard to read.

How about:

To illustrate the above with an example, a pattern sequence `"a+ b"` (one or more `"a"`s followed by a `"b"`) with 

input `"a1", "c", "a2", "b"` will have the following results:

 1. Strict Contiguity: `{a2 b}` because there is `"c"` `"a1"` and `"a2"` so `"a1"` is discarded.

 2. Relaxed Contiguity: `{a1 b}` and `{a1 a2 b}`, as `"c"` will get simply ignored.
 
 3. Non-Deterministic Relaxed Contiguity: `{a1 b}`, `{a2 b}` and `{a1 a2 b}`.

File: docs/dev/libs/cep.md [hunk 11]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
-</div>
-</div>
+ 2. Relaxed Contiguity: which simply ignores non-matching events appearing in-between the matching ones.
+ 
+ 3. Non-Deterministic Relaxed Contiguity: which further relaxes contiguity by also creating alternative
+ matches which ignore also matching events.
 
-Non-strict contiguity means that other events are allowed to occur in-between two matching events.
-A non-strict contiguity pattern state can be created via the `followedBy` or `followedByAny` method.
+To illustrate the above with an example, a pattern sequence `a+ b` (one or more `a`s followed by a `b`) with
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: First I would start with example without looping state, as we silently introduce the internal continuity.

File: docs/dev/libs/cep.md [hunk 12]
```diff
@@ -246,63 +333,118 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
 </div>
-
 </div>
-It is also possible to define a temporal constraint for the pattern to be valid.
-For example, one can define that a pattern should occur within 10 seconds via the `within` method. 
+
+Bear in mind that relaxed contiguity means that only the first succeeding matching event will be matched, while
+non-deterministic relaxed contiguity, multiple matches will be emitted for the same beginning.
+
+Finally, it is also possible to define a temporal constraint for the pattern to be valid.
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: I would emphasize it applies whole pattern sequence. And if multiple within are applied (even for different states/parts) the shortes one is picked up.

**Comment by rmetzger** [comment]
@dawidwys Do you think we can merge the updated documentation today already?
I would like to put out the 1.3 release ann with a link to the CEP docs.

File: docs/dev/libs/cep.md [hunk 13]
```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
+
+<span class="label label-danger">Attention</span> State names **CANNOT** contain the character `:`.
+
+In the remainder, we start by describing how to define [States](#states), before describing how you can 
+combine individual states into [Complex Patterns](#combining-states).
+
+### Individual States
+
+A **State** can be either a *singleton* state, or a *looping* one. Singleton states accept a single event, 
+while looping ones accept more than one. In pattern matching symbols, in the pattern `a b+ c? d` (or `a`,
```
Comment (dawidwys) [inline docs/dev/libs/cep.md]: The sentence starting with "In pattern matchin" was hard to understand for me.

How about?:
In regex style symbols, in the pattern:
 > `a b+ c? d`
   
or in more FLinkCEP style:
 > ( `a`, followed by *one or more* `b`'s, optionally followed by a `c`, followed by a `d`),
  
`a`, `c?`, and `d` are singleton patterns, while `b+` is a looping one (see [Quantifiers](#quantifiers)).

**Comment by dawidwys** [comment]
In fact I managed to finish the review. Generally the docs look great! Congrats on the hard work @kl0u! 

@rmetzger Most of the comments I had were to the formatting/reordering, the only bigger concern is the naming of the State. All in all I think it can be merged today.

**Comment by rmetzger** [comment]
Cool, thx

**Comment by kl0u** [comment]
Thanks a lot @dawidwys for having a detailed look at it! I will integrate your comments and then merge.

**Comment by rmetzger** [comment]
I'll update the release announcement blog post, once the CEP doc has been updated

File: docs/dev/libs/cep.md [hunk 14]
```diff
@@ -168,33 +256,34 @@ start.where(
 </div>
 </div>
 
-<span class="label label-danger">Attention</span> The call to `Context.getEventsForPattern(...)` has to find the 
-elements that belong to the pattern. The cost of this operation can vary, so when implementing your condition, try 
-to minimize the times the method is called.
+<span class="label label-danger">Attention</span> The call to `context.getEventsForPattern(...)` finds all the 
+previously accepted events for a given potential match. The cost of this operation can vary, so when implementing 
+your condition, try to minimize the times the method is called.
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... try to minimize its use.

File: docs/dev/libs/cep.md [hunk 15]
```diff
@@ -168,33 +256,34 @@ start.where(
...
 
-**Simple Conditions:** This type of conditions extend the aforementioned `IterativeCondition` class. They are simple 
-filtering conditions that decide to accept an element or not, based only on properties of the element itself.
+**Simple Conditions:** This type of conditions extend the aforementioned `IterativeCondition` class and decides
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... This type of condition extends the aforementioned `IterativeCondition` class and decides whether ...

File: docs/dev/libs/cep.md [hunk 16]
```diff
@@ -98,48 +128,106 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
-Below is the code for an iterative condition that accepts elements whose name start with "foo" and for which, the sum 
-of the prices of the previously accepted elements for a state named "middle", plus the price of the current event, do 
-not exceed the value of 5.0. Iterative condition can be very powerful, especially in combination with quantifiers, e.g.
-`oneToMany` or `zeroToMany`.
+The pattern API allows you to quickly define complex pattern sequences that you want to extract 
+from your input stream.
+
+Each such complex pattern sequence consists of multiple simple patterns, i.e. patterns looking for 
+individual events with the same properties. These simple patterns are called **states**. A complex pattern 
+can be seen as a graph of such states, where transition from one state to the next happens based on user-specified
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... where transitions from one state to the next occur ...

File: docs/dev/libs/cep.md (hunk 10 above)
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: +1 for the added formatting. I find the explanations hard to parse. How about this:

1. Strict Contiguity: `{a2 b}` -- the `"c"` after `"a1"` causes `"a1"` to be discarded.

2. Relaxed Contiguity: `{a1 b}` and `{a1 a2 b}` -- `"c"` is simply ignored.

3. Non-Deterministic Relaxed Contiguity: `{a1 b}`, `{a2 b}`, and `{a1 a2 b}`.

File: docs/dev/libs/cep.md [hunk 17]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
-A strict contiguity pattern state can be created via the `next` method.
+FlinkCEP supports the following forms of contiguity between events:
+
+ 1. Strict Contiguity: which expects all matching events to appear strictly the one after the other,
+ without any non-matching events in-between.
+
+ 2. Relaxed Contiguity: which simply ignores non-matching events appearing in-between the matching ones.
+ 
+ 3. Non-Deterministic Relaxed Contiguity: which further relaxes contiguity by also creating alternative
+ matches which ignore also matching events.
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: Non-Deterministic Relaxed Contiguity is hard to understand. Maybe this is clearer:

... which further relaxes contiguity, allowing additional matches that ignore some matching events.

File: docs/dev/libs/cep.md [hunk 18]
```diff
@@ -700,26 +995,29 @@ class MyPatternFlatSelectFunction<IN, OUT> implements PatternFlatSelectFunction<
...
 <div data-lang="scala" markdown="1">
-The `select` method takes a selection function as argument, which is called for each matching event sequence.
-It receives a map of string/event pairs of the matched events.
-The string is defined by the name of the state to which the event has been matched.
-The selection function returns exactly one result per call.
+The `select()` method takes a selection function as argument, which is called for each matching event sequence.
+It receives a match in the form of `Map[String, Iterable[IN]]` where the key is the name of each state in your pattern 
+sequence and the value is an Iterable over all accepted events for that state (`IN` is the type of your input elements). 
+The events for a given state are ordered by timestamp. The reason for returning an iterable of accepted events for each 
+state is that when using looping states (e.g. `oneToMany()` and `times()`), more than one events may be accepted for a
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... more than one event may be accepted ...

File: docs/dev/libs/cep.md [hunk 19]
```diff
@@ -700,26 +995,29 @@ class MyPatternFlatSelectFunction<IN, OUT> implements PatternFlatSelectFunction<
...
+    val endEvent = pattern.get("end").get.next
     OUT(startEvent, endEvent)
 }
 {% endhighlight %}
 
-The `flatSelect` method is similar to the `select` method. Their only difference is that the function passed to the `flatSelect` method can return an arbitrary number of results per call.
-In order to do this, the function for `flatSelect` has an additional `Collector` parameter which is used for the element output.
+The `flatSelect` method is similar to the `select` method. Their only difference is that the function passed to the 
+`flatSelect` method can return an arbitrary number of results per call. In order to do this, the function for 
+`flatSelect` has an additional `Collector` parameter which is used for forwarding your output elements downstream.
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... which is used to forward ...

File: docs/dev/libs/cep.md [hunk 20]
```diff
@@ -511,133 +849,81 @@ val start = Pattern.begin[Event]("start")
...
-                <p>Adds a new filter condition which is ORed with an existing filter condition. Only if an event passes the filter condition, it can match the state:</p>
-{% highlight scala %}
-patternState.where(event => ... /* some condition */)
-    .or(event => ... /* alternative condition */)
-{% endhighlight %}
-                    </td>
+                    <td><strong>followedByAny()</strong></td>
+                    <td>
+                        <p>Appends a new pattern state. Other events can occur between a matching event and the previous 
+                        matching event and alternative matches will be presented for every alternative matching event
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: add a comma:

... matching event, and alternative matches will be ...

File: docs/dev/libs/cep.md [hunk 21]
```diff
@@ -341,153 +765,67 @@ Pattern<Event, ?> start = Pattern.<Event>begin("start");
...
 {% endhighlight %}
             </td>
         </tr>
         <tr>
-            <td><strong>Where</strong></td>
+            <td><strong>followedByAny()</strong></td>
             <td>
-                <p>Defines a condition for the current pattern state. Only if an event satisifes the condition, it can match the state:</p>
+                <p>Appends a new pattern state. Other events can occur between a matching event and the previous 
+                matching event and alternative matches will be presented for every alternative matching event
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: add a comma:

... matching event, and alternative matches ...

File: docs/dev/libs/cep.md [hunk 22]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+
 {% endhighlight %}
 </div>
-
 </div>
-It is also possible to define a temporal constraint for the pattern to be valid.
-For example, one can define that a pattern should occur within 10 seconds via the `within` method. 
+
+Bear in mind that relaxed contiguity means that only the first succeeding matching event will be matched, while
+non-deterministic relaxed contiguity, multiple matches will be emitted for the same beginning.
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... will be matched, while with non-deterministic ...

File: docs/dev/libs/cep.md [hunk 23]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+</table>
 </div>
+
 </div>
 
-Non-strict contiguity means that other events are allowed to occur in-between two matching events.
-A non-strict contiguity pattern state can be created via the `followedBy` or `followedByAny` method.
+### Combining States
+
+Now that we have seen how an individual state can look, it is time to see how to combine them into a full pattern sequence.
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: Now that we have seen what an individual state can look like, ...

File: docs/dev/libs/cep.md [hunk 24]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+patternState.where(event => ... /* some condition */)
+    .or(event => ... /* alternative condition */)
+{% endhighlight %}
+                    </td>
+                </tr>
+       <tr>
+           <td><strong>subtype(subClass)</strong></td>
+           <td>
+               <p>Defines a subtype condition for the current pattern state. Only if an event is of this subtype, 
+               it can match the state:</p>
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: You meant to say "Only if an event is of this subtype, can it match the state:", but I recommend

... An event can only match the state if it is of this subtype:

File: docs/dev/libs/cep.md [hunk 25]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
-val strictNext: Pattern[Event, _] = start.next("middle")
+patternState.where(event => ... /* some condition */)
 {% endhighlight %}
+            </td>
+        </tr>
+        <tr>
+            <td><strong>or(condition)</strong></td>
+            <td>
+                <p>Adds a new condition which is ORed with an existing one. Only if an event passes one of the 
+                conditions, it can match the state:</p>
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: An event can only match the state if it passes one of the conditions:

File: docs/dev/libs/cep.md [hunk 26]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+            <th class="text-center">Description</th>
+        </tr>
+    </thead>
+    <tbody>
+      
+        <tr>
+            <td><strong>where(condition)</strong></td>
+            <td>
+              <p>Defines a condition for the current state. Only if an event satisifes the condition, 
+              it can match the state. Multiple consecutive where() clauses lead to their condtions being
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: ... can it match the state.

Or turn the sentence around to be more direct:

An event matches the state only if it satisfies the condition.

File: docs/dev/libs/cep.md [hunk 27]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+    }
+});
+{% endhighlight %}
+            </td>
+        </tr>
+        <tr>
+            <td><strong>or(condition)</strong></td>
+            <td>
+                <p>Adds a new condition which is ORed with an existing one. Only if an event passes one of the 
+                conditions, it can match the state:</p>
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: An event can match the state only if it passes at least one of the conditions.

File: docs/dev/libs/cep.md [hunk 28]
```diff
@@ -246,63 +334,399 @@ pattern.where(event => ... /* some condition */).or(event => ... /* or condition
...
+            <th class="text-left" style="width: 25%">Pattern Operation</th>
+            <th class="text-center">Description</th>
+        </tr>
+    </thead>
+    <tbody>
+       <tr>
+            <td><strong>where(condition)</strong></td>
+            <td>
+                <p>Defines a condition for the current state. Only if an event satisifes the condition, 
+                it can match the state. Multiple consecutive where() clauses lead to their condtions being
```
Comment (alpinegizmo) [inline docs/dev/libs/cep.md]: To match the state, an event must satisfy the condition.

File: docs/dev/libs/cep.md (hunk 2 above)
Comment (kl0u) [inline docs/dev/libs/cep.md]: I agree. The problem is that then we have many back-and-forths. I will try to somehow integrate this comments (probably with a forward reference like "more on contiguity ...").

File: docs/dev/libs/cep.md [hunk 29]
```diff
@@ -98,48 +128,105 @@ val result: DataStream[Alert] = patternStream.select(createAlert(_))
...
-Each pattern consists of multiple stages or what we call states.
-In order to go from one state to the next, the user can specify conditions.
-These conditions can be the contiguity of events or a filter condition on an event.
-
-Each pattern has to start with an initial state:
+The pattern API allows you to quickly define complex pattern sequences that you want to extract 
+from your input stream.
+
+Each such complex pattern sequence consists of multiple simple patterns, i.e. patterns looking for 
+individual events with the same properties. These simple patterns are called **states**. A complex pattern
```
Comment (kl0u) [inline docs/dev/libs/cep.md]: I changed it, so now we have `Pattern` and `Pattern Sequence`.

File: docs/dev/libs/cep.md [hunk 30]
```diff
@@ -1155,5 +1199,9 @@ The changes required are:
 implementing the `FilterFunction` interface.
 
 2. Change your functions provided as arguments to the `select(...)` and `flatSelect(...)` methods to expect a list of
-events associated with each state (`List` in `Java`, `Iterable` in `Scala`). This is because with the addition of
-the looping states, multiple input events can much a single (looping) state.
\ No newline at end of file
+events associated with each pattern (`List` in `Java`, `Iterable` in `Scala`). This is because with the addition of
+the looping patterns, multiple input events can match a single (looping) pattern.
+
+3. The `followedBy()` in Flink 1.1 and 1.2 implied `non-deterministic relaxed contiguity` (see
```
Comment (tzulitai) [inline docs/dev/libs/cep.md]: Looks good! 😄
